go/registry/registry.go

```diff
@@ -0,0 +1,239 @@
+/*
+	Copyright 2022 Loophole Labs
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		   http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+package registry
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/loopholelabs/scalefile/scalefunc"
+	"io"
+	"net/http"
+	"os"
+)
+
+type PullPolicy string
+
+const (
+	AlwaysPullPolicy       PullPolicy = "always"
+	IfNotPresentPullPolicy            = "if-not-present"
+	NeverPullPolicy                   = "never"
+)
+
+var (
+	ErrHashMismatch = errors.New("hash mismatch")
+	ErrNoFunction   = errors.New("function does not exist and pull policy is never")
+)
+
+type Config struct {
+	pullPolicy     PullPolicy
+	cacheDirectory string
+	apiKey         string
+	apiBaseUrl     string
+	organization   string
+}
+
+type Option func(config *Config)
+
+func WithPullPolicy(pullPolicy PullPolicy) Option {
+	return func(config *Config) {
+		config.pullPolicy = pullPolicy
+	}
+}
+
+func WithCacheDirectory(cacheDirectory string) Option {
+	return func(config *Config) {
+		config.cacheDirectory = cacheDirectory
+	}
+}
+
+func WithApiKey(apiKey string) Option {
+	return func(config *Config) {
+		config.apiKey = apiKey
+	}
+}
+
+func WithBaseUrl(apiBaseUrl string) Option {
+	return func(config *Config) {
+		config.apiBaseUrl = apiBaseUrl
+	}
+}
+
+func WithOrganization(organization string) Option {
+	return func(config *Config) {
+		config.organization = organization
+	}
+}
+
+// Create a new runtime for a specific scalefile
+func New(function string, tag string, opts ...Option) (*scalefunc.ScaleFunc, error) {
+	// Default config
+	conf := &Config{
+		pullPolicy:     AlwaysPullPolicy,
+		cacheDirectory: "~/.cache/scale/functions",
```

Comment:
save this in a const or use the homedir lib (https://pkg.go.dev/github.com/mitchellh/go-homedir) to get this?